### PR TITLE
Add missing problem

### PR DESCRIPTION
**Pull request by frcroth**
Feedback welcome

- [x] Add solution notes

### CODE REVIEW

**Comment by frcroth** [comment]
Generally the problem is ok, but it needs to be clear that the output needs to be as small as possible

File: cp22/hpi-fun2/missing/problem.tex [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+
+After they left citadel space, Commander Shepard of the Normandy was attacked by unknown forces and escaped, separating from the rest of the fleet. While the rest of the fleet is still beyond the Kappa Iotta Mass Relay, they have now regrouped and have to decide how to proceed with their mission.
+
+While the ship received a message from the rest of the fleet beyond the Kappa Iotta Mass Relay, it is encrypted. The ship's board computer can decrypt single words, the entire message is not decipherable, since the words have been transmitted together, with the spaces missing.
+
+The decryption code ensures that every word is a palindrome, and every palindrome is a valid word. Can you help Shepard and the Normandy's crew to decipher the message by determining how many words are at least contained?
+
+\paragraph*{Input}
+
+A string $s$ of lowercase letters of length $n$, $1 \geq n \geq 2500$.
```
Comment (schmidtjonas) [inline cp22/hpi-fun2/missing/problem.tex]:
```suggestion
A string $s$ of lowercase letters of length $n$, $1 \le n \le 2500$.
```

File: cp22/hpi-fun2/missing/executables/generator.py [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+some_text = "sqddbydscraqgvzfquizlmdopleiqtceyemhuyotzj"
+write_testcase("single_letter", "Only one letter", "x")
+write_testcase("long_palindrome", "A long palindrome", some_text + "".join(reversed(some_text)))
+write_testcase("no_palindromes", "A testcase without palindromes", "azbcdefgzhijklmznopqrsztuvw")
+long_text = ''.join(random.choice(string.ascii_lowercase) for _ in range(int(max_length/2)))
+write_testcase("very_long_palindrome", "A very long palindrome", long_text + "".join(reversed(long_text)))
+write_testcase("many_palindromes", "Many similar palindromes", "".join(random.choice(["anna", "otto", "iggi"]) for _ in range(int((max_length-1)/4))))
+write_testcase("random_small", "A small random testcase",''.join(random.choice(string.ascii_lowercase) for _ in range(20)))
+write_testcase("random_large_1", "A large random testcase",''.join(random.choice(string.ascii_lowercase) for _ in range(max_length)))
+write_testcase("random_large_2", "A large random testcase",''.join(random.choice(string.ascii_lowercase) for _ in range(max_length)))
```
Comment (schmidtjonas) [inline cp22/hpi-fun2/missing/executables/generator.py]: Maybe add some more random cases.
```suggestion
for i in range(5):
    write_testcase(f"random_small_{i}", "A small random testcase",''.join(random.choice(string.ascii_lowercase) for _ in range(20)))
    write_testcase("random_large_{i}", "A large random testcase",''.join(random.choice(string.ascii_lowercase) for _ in range(max_length)))
```

File: cp22/hpi-fun2/missing/problem.json [hunk 3]
```diff
@@ -0,0 +1,11 @@
+{
+  "based_on": {
+    "data": [
+      "https://www.geeksforgeeks.org/palindrome-partitioning-dp-17/"
+    ],
+    "type": "other"
+  },
+  "description": "Palindromic cuts",
+  "difficulty": 3,
+  "tags": []
```
Comment (schmidtjonas) [inline cp22/hpi-fun2/missing/problem.json]:
```suggestion
  "tags": [
    "dp",
    "strings"
  ]
```

File: cp22/hpi-fun2/missing/executables/generator.py [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+
+max_length = 2500
+
+write_testcase("sample1", "A simple testcase", "allaoggo")
+write_testcase("sample2", "A simple testcase", "banana")
+write_testcase("sample3", "A simple testcase", "masseffect")
+some_text = "sqddbydscraqgvzfquizlmdopleiqtceyemhuyotzj"
+write_testcase("single_letter", "Only one letter", "x")
+write_testcase("long_palindrome", "A long palindrome", some_text + "".join(reversed(some_text)))
+write_testcase("no_palindromes", "A testcase without palindromes", "azbcdefgzhijklmznopqrsztuvw")
```
Comment (schmidtjonas) [inline cp22/hpi-fun2/missing/executables/generator.py:21]: Maybe add a dedicated "anti-greedy" case, which punishes you for taking the longest palindrome possible at every position like `abbacddcbbae`. Greedy takes abba, cddc, bb, a, e. Optimal is a, bbacddcbba, e.

**Review by schmidtjonas — COMMENTED** [review]
nice, I don't think there is much left to do besides adding solution notes.
The time limit and input bounds look good. The pdf and test cases look good too except for my comments.

File: cp22/hpi-fun2/missing/executables/generator.py [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+long_text = ''.join(random.choice(string.ascii_lowercase) for _ in range(int(max_length/2)))
+write_testcase("very_long_palindrome", "A very long palindrome", long_text + "".join(reversed(long_text)))
+write_testcase("anti_greedy", "A testcase that has a different result for a greedy algo", "abbacddccddcabba" * int(1500/16))
```
Comment (schmidtjonas) [inline cp22/hpi-fun2/missing/executables/generator.py]: Your anti greedy has solution 1. Maybe just add the testcase I proposed 😅

File: cp22/hpi-fun2/missing/executables/generator.py [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+write_testcase("many_palindromes", "Many similar palindromes", "".join(random.choice(["anna", "otto", "iggi"]) for _ in range(int((max_length-1)/4))))
+for i in range(5):
+    write_testcase(f"random_small_{i}", "A small random testcase",''.join(random.choice(string.ascii_lowercase) for _ in range(20)))
+    write_testcase("random_large_{i}", "A large random testcase",''.join(random.choice(string.ascii_lowercase) for _ in range(max_length)))
```
Comment (schmidtjonas) [inline cp22/hpi-fun2/missing/executables/generator.py]: Oops, I forgot an `f`
```suggestion
    write_testcase(f"random_large_{i}", "A large random testcase",''.join(random.choice(string.ascii_lowercase) for _ in range(max_length)))
```

**Comment by schmidtjonas** [comment]
looks good to me

**Pull request merged**